Add agentic code generation workload well-lit path (#1727)
llm-d/llm-d: guides/agentic-serving/agentic-code-generation-tpu.md

| ------------------ | ------------------------------------------------------------------------------------------ |
| Model              | [Qwen/Qwen3-Coder-480B-A35B-Instruct-FP8](https://huggingface.co/Qwen/Qwen3-Coder-480B-A35B-Instruct-FP8) |
| Replicas           | 8                                                                                          |
| Accelerator        | Google TPU v7 (tpu7x)                                                                      |
| Topology           | 2x2x1                                                                                      |
| TP size / EP size  | TP=8, EP enabled                                                                           |

### Supported Hardware Backends

| Backend             | Directory                      | Notes                       |
| ------------------- | ------------------------------ | --------------------------- |
| Google TPU (vLLM)   | `modelserver/tpu/vllm/`        | TPU v7 / tpu7x 2x2x1 (nightly) |

## Prerequisites

- Installed proper client tools (kubectl, helm).
- Set the following environment variables:
  ```bash
  export GAIE_VERSION=v1.5.0
  export GUIDE_NAME="agentic-workloads"

>   export NAMESPACE=llm-d-agentic-workloads

Reviewer: TO be consistent, pls also add HF_TOKEN and REPO_ROOT, see optimized baseline for an example.

Author: Done

# Agentic Code Generation Guide

## Overview

This guide deploys the optimal llm-d configuration for agentic workloads using agentic code generation as the example workload. The configuration includes multiple llm-d optimizations in terms of routing and KV cache management:
- **Prefix-aware routing** to optimize prefix cache reuse (via `prefix-cache-scorer`)
- **KV cache offloading** to CPU DRAM to handle multi-turn conversations with long contexts (via KV offloading connector)

> - **Load balancing** to prevent replica hotspotting from bursty request patterns (via `queue-scorer` and `kv-cache-utilization-scorer`)

Reviewer 2: are we not using the prefix-cache filter @kaushikmitr ?

Reviewer 3: Assuming my suggestions are accepted below, the use of the filter should be within the router config + benchmarking results, not here.

Author: Included @vMaroon's suggestions to remove the exact scorer and other routing config used from the guide.

@ahg-g on the prefill-cache filter, this configuration follows optimized-baseline and the benchmark numbers are based on that too. Can we do any additional tuning after merging this? This'd help unblock the GPU agentic guide as well.

Reviewer 2: Sounds good

# Agentic Code Generation on TPU

## Overview

This guide deploys the optimal llm-d configuration for agentic workloads using agentic code generation as the example workload. The configuration includes multiple llm-d optimizations in terms of routing and KV cache management:

> - **Prefix-aware routing** to optimize prefix cache reuse (via `prefix-cache-scorer`)

Reviewer 3: ```suggestion
- **Prefix-aware routing** to optimize prefix cache reuse
```

I think we should keep it high level

- **KV cache offloading** to CPU DRAM to handle multi-turn conversations with long contexts (via KV offloading connector)

> - **Load balancing** to prevent replica hotspotting from bursty request patterns (via `queue-scorer` and `kv-cache-utilization-scorer`)

Reviewer 3: ```suggestion
- **Load balancing** to increase cluster-wide accelerator utilization and prevent hot-spotting from bursty request patterns
```

I think we should keep it high level

> ## Workload Profile

Reviewer 3: This should be moved under benchmarking

Author: Done

| ------------------ | ------------------------------------------------------------------------------------------ |
| Model              | [Qwen/Qwen3-Coder-480B-A35B-Instruct-FP8](https://huggingface.co/Qwen/Qwen3-Coder-480B-A35B-Instruct-FP8) |
| Replicas           | 8                                                                                          |
| Accelerator        | Google TPU v7 (tpu7x)                                                                      |
| Topology           | 2x2x1                                                                                      |
| TP size / EP size  | TP=8, EP enabled                                                                           |

### Supported Hardware Backends

| Backend             | Directory                      | Notes                       |
| ------------------- | ------------------------------ | --------------------------- |
| Google TPU (vLLM)   | `modelserver/tpu/vllm/`        | TPU v7 / tpu7x 2x2x1 (nightly) |

## Prerequisites

- Installed proper client tools (kubectl, helm).
- Set the following environment variables:
  ```bash
  export GAIE_VERSION=v1.5.0

>   export GUIDE_NAME="agentic-workloads"

Reviewer 3: ```suggestion
  export GUIDE_NAME="agentic-serving"
```

>   export NAMESPACE=llm-d-agentic-workloads

Reviewer 3: ```suggestion
  export NAMESPACE=llm-d-agentic-serving
```

### 2. Download the Workload Template

```bash
curl -LJO "https://raw.githubusercontent.com/llm-d/llm-d/main/guides/${GUIDE_NAME}/benchmark-templates/guide.yaml"
```

### 3. Execute Benchmark

```bash
export IP=$(kubectl get service ${GUIDE_NAME}-epp -n ${NAMESPACE} -o jsonpath='{.spec.clusterIP}')
envsubst < guide.yaml > config.yaml
./run_only.sh -c config.yaml -o ./results
```

## Benchmark Results

The results below are with 8 replicas of TPU v7x (2x2x1) on the benchmark workload described above.

### Summary with 40 concurrent coding sessions:
| Metric | k8s Service | llm-d-optimized | Δ Improvement |
| :--- | :--- | :--- | :--- |
| **TTFT P50 (ms)** | 17391 | 1314 | ⬇️ 92.4% |
| **Input tokens / sec** | 36987 | 78065 | ⬆️ 111.1% |
| **Output tokens / sec** | 436.8 | 888.5 | ⬆️ 103.4% |

### Latency Profiles:

<p float="left">
  <img src="./benchmark-results/latency_vs_throughput.png" width="32%" alt="Latency vs Throughput" />
  <img src="./benchmark-results/throughput_vs_concurrency.png" width="32%" alt="Throughput vs Concurrency" />
  <img src="./benchmark-results/ttft_vs_concurrency.png" width="32%" alt="TTFT vs Concurrency" />
</p>

> **Note**: As of June 2026 we are actively working to improve:

Reviewer 3: ```suggestion
**Note**: As of June 2026 we are actively working to improve for TPU deployments:
```
